scalyr_agent/monitor_utils/k8s.py

```diff
@@ -151,37 +187,24 @@ def __init__( self, name='', namespace='', labels={} ):
 
 class _K8sCache( object ):
     """
-        A lazily updated cache of objects from a k8s api query
+        A cached store of objects from a k8s api query
 
         This is a private class to this module.  See KubernetesCache which instantiates
         instances of _K8sCache for querying different k8s API objects.
 
-        A full update of cached data occurs whenever the cache is queried and
-        `cache_expiry_secs` have passed since the last full update.
-
-        If a cache miss occurs e.g. objects were created since the last cache update but
-        that is not yet reflected in the cached data, individual queries to the
-        api are made that get the required data only, and no more.
-
-        If too many cache misses occur within `cache_miss_interval` (e.g. a large
-        number of objects were created since the last full cache update) this will also
-        trigger a full update of the cache.
-
         This abstraction is thread-safe-ish, assuming objects returned
         from querying the cache are never written to.
     """
 
-    def __init__( self, logger, processor, object_type, cache_expiry_secs=120, max_cache_misses=20, cache_miss_interval=10 ):
+    def __init__( self, logger, k8s, processor, object_type, filter=None, perform_full_updates=True ):
         """
             Initialises a Kubernees Cache
             @param: logger - a Scalyr logger
+            @param: k8s - a KubernetesApi object
             @param: processor - a _K8sProcessor object for querying/processing the k8s api
             @param: object_type - a string containing a textual name of the objects being cached, for use in log messages
-            @param: cache_expiry_secs - the number of seconds to wait before doing a full update of data from the k8s api
-            @param: max_cache_misses - the maximum of number of cache misses that can occur in
-                                       `cache_miss_interval` before doing a full update of data from the k8s api
-            @param: cache_miss_interval - the number of seconds that `max_cache_misses` can occur in before a full
-                                          update of the cache occurs
+            @param: filter - a k8s filter string or none if no filtering is to be done
+            @param: perform_full_updates - Boolean.  If False no attempts will be made to fully update the cache (only single items will be cached)
```

Comment:
I thought we weren't ever going to do full updates anymore after the initial cache sync.  Is that not true?

Reply:
Only for controller objects.  The pod objects still need to be continually refreshed in case their annotations change.  

We could not do a full refresh, but then then we'd be re-querying each pod anyway, but instead of 1 query returning 100 pods, we'd have 100 queries returning a single pod.

I figure the former is better than the latter and since switching to ujson the overhead for a full refresh, even for a large number of pods appears to be minimal.

Reply:
OK.. so the controller objects, we will query them one-by-one when they expire, correct? That way if the labels on a deployment/daemonset/etc changes, we will eventually notice the label changes.  I just want to double check this.

Reply:
For the controller objects, we always query them one by one, and cache the results.

We likely won't notice labels that change, because controller objects only expire from the cache *if they haven't been accessed* in `k8s_cache_purge_secs`.

The purging is really only so that we no longer hold controller objects in memory if they're not being used, e.g. after a deployment has been deleted, there will be no longer any need to keep that controller around in memory, and because we no longer need to query it it will eventually be purged.

Therefore, as long as a controller object is being queried (e.g. every time through the gather_sample loop and the container_checker loop) it will be marked as alive, and therefore won't expire, and therefore won't be requeried.

It's not to difficult to change this behaviour if needed.

Reply:
Hmm.. yes.. this one does worry me.  Did we decide at one point to not worry about the label changes?  I can't remember.  But, I can see someone having long running deployments (where they always update using a rolling update).  I would think labels could be updated and we should notice those changes.

We need to get this change in.  If this work is going to take a bit of time, then let's just get this change checked in and have you follow up with another PR if you agree we should address this.